### ISSUE TITLE
Test checks for fails in unpredictable order

### ISSUE THREAD

**Issue by Tux** [petdance/html-lint#40]
<pre>t/12-html_fragment_ok.t ......... 1/16
#   Failed test 'Error matches string'
#   at t/LintTest.pl line 36.
#          got: 'Set #1 (6:1) <html> tag is required'
#     expected: 'Set #1 (6:1) <head> tag is required'

#   Failed test 'Error matches string'
#   at t/LintTest.pl line 36.
#          got: 'Set #1 (6:1) <head> tag is required'
#     expected: 'Set #1 (6:1) <html> tag is required'
# Looks like you failed 2 tests of 16.
t/12-html_fragment_ok.t ......... Dubious, test returned 2 (wstat 512, 0x200)
Failed 2/16 subtests</pre>
<p>Making the order of the tests in the test unambiguous:</p>
<pre>
diff -purdw HTML-Lint-2.22/t/LintTest.pl HTML-Lint-2.22-r67MGf/t/LintTest.pl
--- HTML-Lint-2.22/t/LintTest.pl        2015-04-02 23:50:54.000000000 +0200
+++ HTML-Lint-2.22-r67MGf/t/LintTest.pl 2015-05-06 14:53:09.333360048 +0200
@@ -18,7 +18,14 @@ sub checkit {
         $lint->eof;
     }

-    my @errors = $lint->errors();
+    my @errors = sort {
+       $a->{_file}    cmp $b->{_file}    ||
+       $a->{_line}    <=> $b->{_line}    ||
+       $a->{_column}  <=> $b->{_column}  ||
+       $a->{_type}    <=> $b->{_type}    ||
+       $a->{_errcode} cmp $b->{_errcode} ||
+       $a->{_errtext} cmp $b->{_errtext}
+       } $lint->errors();
     is( scalar @errors, scalar @expected, 'Right # of errors' );

     while ( @errors && @expected ) {
diff -purdw HTML-Lint-2.22/t/text-use-entity.t HTML-Lint-2.22-r67MGf/t/text-use-entity.t
--- HTML-Lint-2.22/t/text-use-entity.t  2015-03-31 20:12:11.000000000 +0200
+++ HTML-Lint-2.22-r67MGf/t/text-use-entity.t   2015-05-06 14:52:56.382228141 +0200
@@ -5,9 +5,9 @@ require 't/LintTest.pl';
 checkit( [
     [ 'text-use-entity'      => qr/Character "\\x0B" should be written as &#11;/ ],
     [ 'text-use-entity'      => qr/Character "\\xF1" should be written as &ntilde;/ ],
+    [ 'text-unclosed-entity' => qr/Entity &ouml; is missing its closing semicolon/ ],
     [ 'text-use-entity'      => qr/Character "&" should be written as &amp;/ ],
     [ 'text-use-entity'      => qr/Character "&" should be written as &amp;/ ],
-    [ 'text-unclosed-entity' => qr/Entity &ouml; is missing its closing semicolon/ ],
 ], [<DATA>] );

 __DATA__
</pre>

<p>Causes all the tests to pass</p>

**Comment by KES777** [comment]
t/12-html_fragment_ok.t ......... 1/16 
#   Failed test 'Error matches string'
#   at t/LintTest.pl line 36.
#          got: 'Set #1 (6:1) <title> tag is required'
#     expected: 'Set #1 (6:1) <html> tag is required'

#   Failed test 'Error matches string'
#   at t/LintTest.pl line 36.
#          got: 'Set #1 (6:1) <html> tag is required'
#     expected: 'Set #1 (6:1) <title> tag is required'
# Looks like you failed 2 tests of 16.
t/12-html_fragment_ok.t ......... Dubious, test returned 2 (wstat 512, 0x200)
Failed 2/16 subtests 
t/20-error-types-export.t ....... ok

**Comment by veryrusty** [comment]
Having just ran into this; it has all the hallmarks of hash randomisation. 

Sorting the errors [here](https://github.com/petdance/html-lint/blob/dev/lib/HTML/Lint/Parser.pm#L257) seems a simple fix, but I'm not sure it there would be any other effects from doing so. Thoughts @petdance ?

**Comment by asan13** [comment]
Hi! Also faced with this. The reason seems [here](https://github.com/petdance/html-lint/blob/dev/lib/HTML/Lint/Parser.pm#L105):

```perl
sub _end_document {
    my ($self,$line,$column) = @_;

    for my $tag ( sort keys %isRequired ) {
        if ( !$self->{_first_seen}->{$tag} ) {
            $self->gripe( 'doc-tag-required', tag => $tag );
        }
    }

    return;
}
```

**Comment by petdance** [comment]
I'm going to look at this tonight.  Will need some tests from the people who have it failing for them if possible.

**Comment by veryrusty** [comment]
@petdance not sure how to offer further tests. The existing test `t/12-html_fragment_ok.t` fails randomly, hence suspecting hash randomization is the root cause.

I use Docker to provide clean environments for testing. This fails ~ 50% or the time
```
docker run --rm -ti perl:5.24 cpanm HTML::Lint
```

**Comment by petdance** [comment]
This has already been fixed on the dev branch.

**Comment by petdance** [comment]
@veryrusty If you had a chance in the next hour to try the dev branch on your machine, that would be nice, but it's not really necessary.  I've got 5.24.0 I'm checking against myself.
